File: android/app/src/main/java/de/phytech/logbot/data/LogbotApi.kt
/**
 * Datei:        LogbotApi.kt
 * Projekt:      Logbot
 * Paket:        de.phytech.logbot.data
 *
 * Beschreibung:
 * Schmaler HTTP-Client fuer die REST-Schnittstelle des Logbot-Servers.
 *
 * Bewusst ohne Bibliothek: Es sind eine Handvoll GET-Aufrufe gegen genau eine
 * Instanz. HttpURLConnection und org.json liegen im System, das spart rund
 * 1 MB APK und eine Abhaengigkeit, die gepflegt werden muesste.
 *
 * Nebenlaeufigkeit: Aufrufe laufen in einem kleinen Thread-Pool, die Antwort
 * kommt auf dem Hauptthread zurueck. Callbacks werden verworfen, sobald der
 * Aufrufer `cancelled` meldet - so schreibt keine Antwort mehr in eine
 * Oberflaeche, die es nicht mehr gibt.
 */
package de.phytech.logbot.data

import android.net.Uri
import android.os.Handler
import android.os.Looper
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors

/** Ergebnis eines Aufrufs: entweder Nutzdaten oder ein eingeordneter Fehler. */
sealed class ApiResult<out T> {
    data class Ok<T>(val value: T) : ApiResult<T>()
    data class Err(val error: ApiError) : ApiResult<Nothing>()
}

/**
 * Fehler mit Einordnung statt nacktem Stacktrace.
 *
 * `status` ist der HTTP-Code, 0 wenn die Verbindung gar nicht zustande kam.
 */
data class ApiError(val status: Int, val message: String) {
    /** 401/403: Token abgelaufen oder zurueckgezogen - die App muss neu einrichten. */
    val unauthorized: Boolean get() = status == 401 || status == 403
    /** 404: Endpunkt gibt es auf diesem Serverstand noch nicht. Kein Fehler des Nutzers. */
    val notImplemented: Boolean get() = status == 404
    val offline: Boolean get() = status == 0
}

class LogbotApi(baseUrl: String, private val token: String) {

    private val base = baseUrl.trimEnd('/')

    // --- Endpunkte ----------------------------------------------------------

    fun health(done: (ApiResult<ServerHealth>) -> Unit) =
        get("/api/health/detailed", done) { ServerHealth.from(it) }

    fun filterOptions(done: (ApiResult<FilterOptions>) -> Unit) =
        get("/api/logs/filter-options", done) { FilterOptions.from(it) }

    fun logs(query: LogQuery, done: (ApiResult<LogPage>) -> Unit) =
        get(query.toPath(), done) { LogPage.from(it) }

    /** Einzelner Eintrag samt Rohtext - erst beim Aufklappen der Detailansicht. */
    fun logDetail(id: Long, done: (ApiResult<LogDetail>) -> Unit) =
        get("/api/logs/$id", done) { LogDetail.from(it) }

    /** Optionaler Endpunkt, siehe docs/SERVER-API.md. 404 = Server kann es noch nicht. */
    fun mailStatus(done: (ApiResult<MailStatus>) -> Unit) =
        get("/api/mail/status", done) { MailStatus.from(it) }

    /** Optionaler Endpunkt: stoesst eine Passwort-Reset-Mail ueber Postfix an. */
    fun requestPasswordReset(login: String, done: (ApiResult<String>) -> Unit) {
        val body = JSONObject().put("login", login).toString()
        post("/api/mail/password-reset", body, done) {
            it.optStringOrEmpty("message").ifEmpty { "Reset-Mail wurde in die Warteschlange gelegt." }
        }
    }

    // --- Innereien ----------------------------------------------------------

    private fun <T> get(path: String, done: (ApiResult<T>) -> Unit, parse: (JSONObject) -> T) =
        request("GET", path, null, done, parse)

    private fun <T> post(
        path: String,
        body: String,
        done: (ApiResult<T>) -> Unit,
        parse: (JSONObject) -> T
    ) = request("POST", path, body, done, parse)

    private fun <T> request(
        method: String,
        path: String,
        body: String?,
        done: (ApiResult<T>) -> Unit,
        parse: (JSONObject) -> T
    ) {
        pool.execute {
            val result: ApiResult<T> = try {
                ApiResult.Ok(parse(JSONObject(execute(method, path, body))))
            } catch (e: ApiException) {
                ApiResult.Err(ApiError(e.status, e.message ?: "Unbekannter Fehler"))
            } catch (e: IOException) {
                ApiResult.Err(ApiError(0, e.message ?: "Server nicht erreichbar"))
            } catch (e: Exception) {
                ApiResult.Err(ApiError(-1, "Antwort nicht lesbar: ${e.message}"))
            }
            main.post { done(result) }
        }
    }

    private fun execute(method: String, path: String, body: String?): String {
        val connection = URL(base + path).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = method
            connection.connectTimeout = CONNECT_TIMEOUT_MS
            connection.readTimeout = READ_TIMEOUT_MS
            connection.setRequestProperty("Authorization", "Bearer $token")
            connection.setRequestProperty("Accept", "application/json")
            if (body != null) {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            }

            val status = connection.responseCode
            if (status in 200..299) {
                return connection.inputStream.bufferedReader().use { it.readText() }
            }

            // Fehlertext des Servers mitnehmen, aber nur wenn er kurz und
            // lesbar ist. FastAPI liefert {"detail": "..."} - alles andere
            // (HTML-Fehlerseiten eines Reverse-Proxy) wuerde nur verwirren.
            val detail = runCatching {
                val raw = connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
                JSONObject(raw).optStringOrEmpty("detail")
            }.getOrDefault("")
            throw ApiException(status, detail.ifEmpty { describe(status) })
        } finally {
            connection.disconnect()
        }
    }

    private fun describe(status: Int): String = when (status) {
        401, 403 -> "Zugang abgelehnt - Token ungueltig oder abgelaufen"
        404 -> "Dieser Serverstand kennt den Endpunkt noch nicht"
        429 -> "Zu viele Anfragen - kurz warten"
        in 500..599 -> "Der Server meldet einen internen Fehler ($status)"
        else -> "Unerwartete Antwort ($status)"
    }

    private class ApiException(val status: Int, message: String) : Exception(message)

    companion object {
        private const val CONNECT_TIMEOUT_MS = 8_000
        private const val READ_TIMEOUT_MS = 20_000

        // Drei Threads reichen: Status, Logs und Mail koennen parallel laufen,
        // mehr gleichzeitige Anfragen stellt die App nie.
        private val pool = Executors.newFixedThreadPool(3)
        private val main = Handler(Looper.getMainLooper())

        /** Baut den Client aus den gespeicherten Zugangsdaten, oder null. */
        fun from(url: String?, token: String?): LogbotApi? =
            if (url.isNullOrBlank() || token.isNullOrBlank()) null else LogbotApi(url, token)
    }
}

/**
 * Filter der Log-Ansicht.
 *
 * Leere Werte werden weggelassen - der Server behandelt einen gesetzten,
 * leeren Parameter sonst als Suche nach dem leeren String.
 */
data class LogQuery(
    val page: Int = 1,
    val pageSize: Int = 50,
    val search: String = "",
    val hostname: String = "",
    val minSeverity: String = "",
    val category: String = ""
) {
    fun toPath(): String {
        val builder = Uri.parse("/api/logs").buildUpon()
            .appendQueryParameter("page", page.toString())
            .appendQueryParameter("page_size", pageSize.toString())
        if (search.isNotBlank()) builder.appendQueryParameter("search", search.trim())
        if (hostname.isNotBlank()) builder.appendQueryParameter("hostname", hostname.trim())
        if (minSeverity.isNotBlank()) builder.appendQueryParameter("min_severity", minSeverity)
        if (category.isNotBlank()) builder.appendQueryParameter("category", category)
        return builder.build().toString()
    }

    /** True, wenn die Liste gerade gefiltert ist - fuer den "Filter zuruecksetzen"-Knopf. */
    val filtered: Boolean
        get() = search.isNotBlank() || hostname.isNotBlank() ||
            minSeverity.isNotBlank() || category.isNotBlank()
}

File: ios/Logbot/LogbotApi.swift
//
//  LogbotApi.swift
//  Logbot
//
//  REST-Zugriff auf den Logbot-Server. Gleiche Endpunkte wie die
//  Android-Fassung, gleiche Einordnung der Fehler.
//
//  Welche Endpunkte Pflicht und welche optional sind: docs/SERVER-API.md
//

import Foundation

// MARK: - Fehler

enum ApiError: LocalizedError {
    case offline(String)
    case unauthorized
    case notImplemented
    case server(Int, String)
    case malformed

    var errorDescription: String? {
        switch self {
        case .offline(let detail):
            return "Server nicht erreichbar: \(detail)"
        case .unauthorized:
            return "Zugang abgelehnt – Token ungültig oder abgelaufen"
        case .notImplemented:
            return "Dieser Serverstand kennt den Endpunkt noch nicht"
        case .server(let code, let detail):
            return detail.isEmpty ? "Server meldet Fehler \(code)" : detail
        case .malformed:
            return "Antwort des Servers war nicht lesbar"
        }
    }
}

// MARK: - Datentypen

struct ServerHealth: Decodable {
    let status: String
    let version: String
    let uptimeSeconds: Double
    let cpuPercent: Double
    let memoryPercent: Double
    let diskPercent: Double
    let databaseConnected: Bool
    let logsTotal: Int
    let logsLast24h: Int
    let agentsTotal: Int
    let agentsOnline: Int

    var healthy: Bool { status.lowercased() == "healthy" }

    enum CodingKeys: String, CodingKey {
        case status, version
        case uptimeSeconds = "uptime_seconds"
        case cpuPercent = "cpu_percent"
        case memoryPercent = "memory_percent"
        case diskPercent = "disk_percent"
        case databaseConnected = "database_connected"
        case logsTotal = "logs_total"
        case logsLast24h = "logs_last_24h"
        case agentsTotal = "agents_total"
        case agentsOnline = "agents_online"
    }
}

struct LogEntry: Decodable, Identifiable, Hashable {
    let id: Int
    let hostname: String?
    let ipAddress: String?
    let timestamp: String
    let level: String?
    let source: String?
    let message: String?

    enum CodingKeys: String, CodingKey {
        case id, hostname, timestamp, level, source, message
        case ipAddress = "ip_address"
    }

    var severity: Severity { Severity.of(level ?? "") }
}

struct LogPage: Decodable {
    let items: [LogEntry]
    let total: Int
    let page: Int
    let pageSize: Int

    var hasMore: Bool { page * pageSize < total }

    enum CodingKeys: String, CodingKey {
        case items, total, page
        case pageSize = "page_size"
    }
}

struct FilterOption: Decodable, Hashable, Identifiable {
    let key: String
    let label: String
    var id: String { key }
}

struct FilterOptions: Decodable {
    let severities: [FilterOption]
    let categories: [FilterOption]

    static let empty = FilterOptions(severities: [], categories: [])
}

struct MailStatus: Decodable {
    let postfixRunning: Bool
    let queueLength: Int
    let deferredLength: Int
    let lastError: String?

    enum CodingKeys: String, CodingKey {
        case postfixRunning = "postfix_running"
        case queueLength = "queue_length"
        case deferredLength = "deferred_length"
        case lastError = "last_error"
    }
}

struct MessageResponse: Decodable {
    let message: String?
}

/// FastAPI meldet Fehler als {"detail": "..."}. Ist `detail` etwas anderes
/// als ein String (etwa eine Liste bei Validierungsfehlern), schlaegt das
/// Dekodieren fehl und der Aufrufer nimmt seinen eigenen Text.
private struct ErrorBody: Decodable {
    let detail: String?
}

/// Schweregrad-Gruppen. Der Server kennt viele Schreibweisen, die Anzeige vier.
enum Severity {
    case error, warn, info, debug

    static func of(_ level: String) -> Severity {
        switch level.lowercased() {
        case "emerg", "emergency", "alert", "crit", "critical", "err", "error", "fatal":
            return .error
        case "warn", "warning", "notice":
            return .warn
        case "debug", "trace":
            return .debug
        default:
            return .info
        }
    }
}

/// Filter der Logansicht. Leere Werte werden weggelassen.
struct LogQuery: Equatable {
    var page = 1
    var pageSize = 50
    var search = ""
    var minSeverity = ""
    var category = ""

    var filtered: Bool { !search.isEmpty || !minSeverity.isEmpty || !category.isEmpty }

    var queryItems: [URLQueryItem] {
        var items = [
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "page_size", value: String(pageSize))
        ]
        if !search.isEmpty { items.append(URLQueryItem(name: "search", value: search)) }
        if !minSeverity.isEmpty { items.append(URLQueryItem(name: "min_severity", value: minSeverity)) }
        if !category.isEmpty { items.append(URLQueryItem(name: "category", value: category)) }
        return items
    }
}

// MARK: - Client

struct LogbotApi {

    let baseURL: String
    let token: String

    init?(baseURL: String?, token: String?) {
        guard let baseURL, let token, !baseURL.isEmpty, !token.isEmpty else { return nil }
        self.baseURL = baseURL
        self.token = token
    }

    func health() async throws -> ServerHealth {
        try await get("/api/health/detailed")
    }

    func filterOptions() async throws -> FilterOptions {
        try await get("/api/logs/filter-options")
    }

    func logs(_ query: LogQuery) async throws -> LogPage {
        try await get("/api/logs", query: query.queryItems)
    }

    /// Optionaler Endpunkt: aeltere Serverstaende antworten mit 404.
    func mailStatus() async throws -> MailStatus {
        try await get("/api/mail/status")
    }

    /// Optionaler Endpunkt: stoesst eine Reset-Mail ueber Postfix an.
    func requestPasswordReset(login: String) async throws -> String {
        let body = try JSONSerialization.data(withJSONObject: ["login": login])
        let response: MessageResponse = try await send("POST", "/api/mail/password-reset", body: body)
        return response.message ?? "Reset-Mail wurde in die Warteschlange gelegt."
    }

    // MARK: Innereien

    private func get<T: Decodable>(_ path: String, query: [URLQueryItem] = []) async throws -> T {
        try await send("GET", path, query: query, body: nil)
    }

    private func send<T: Decodable>(
        _ method: String,
        _ path: String,
        query: [URLQueryItem] = [],
        body: Data? = nil
    ) async throws -> T {
        guard var components = URLComponents(string: baseURL + path) else { throw ApiError.malformed }
        if !query.isEmpty { components.queryItems = query }
        guard let url = components.url else { throw ApiError.malformed }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 20
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw ApiError.offline(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else { throw ApiError.malformed }
        switch http.statusCode {
        case 200...299:
            do {
                return try JSONDecoder().decode(T.self, from: data)
            } catch {
                throw ApiError.malformed
            }
        case 401, 403:
            throw ApiError.unauthorized
        case 404:
            throw ApiError.notImplemented
        default:
            // FastAPI liefert {"detail": "..."}. Alles andere (etwa die
            // HTML-Fehlerseite eines Reverse-Proxy) wuerde nur verwirren.
            let detail = (try? JSONDecoder().decode(ErrorBody.self, from: data))?.detail ?? ""
            throw ApiError.server(http.statusCode, detail)
        }
    }
}
